include/emp/Evolve/Systematics.hpp

```diff
@@ -1810,31 +1781,645 @@ namespace emp {
+        // std::cout << "Alive point" << " depth: " << depth << " divisor: " << divisor << std::endl;
+        total += depth/divisor;
+        depth = 0;
+        divisor = test_taxon->GetTotalOffspring() + 1;
+      } else if (test_taxon->GetNumOff() > 1) {
+        // This is a branch point. We need to add the things on the other branch to the divisor..
+        // std::cout << "Branch point" << " depth: " << depth << " divisor: " << divisor << std::endl;
+        total += depth/divisor;
+        depth = 0;
+        divisor = test_taxon->GetTotalOffspring();
+      }
+
+      test_taxon = test_taxon->GetParent();
+    }
+
+    return -1;
+  }
+
+  template <typename ORG, typename ORG_INFO, typename DATA_STRUCT>
+  int Systematics<ORG, ORG_INFO, DATA_STRUCT>::GetBranchesToRoot(Ptr<taxon_t> tax) const {
+    GetMRCA();
+
+    int depth = 0;
+    Ptr<taxon_t> test_taxon = tax->GetParent();
+    while (test_taxon) {
+      if (test_taxon == mrca || !test_taxon) {
+        return depth;
+      } else if (test_taxon->GetNumOff() > 1) {
+        depth++;
+      }
+      test_taxon = test_taxon->GetParent();
+    }
+    return depth;
+  }
+
+  template <typename ORG, typename ORG_INFO, typename DATA_STRUCT>
+  int Systematics<ORG, ORG_INFO, DATA_STRUCT>::GetDistanceToRoot(Ptr<taxon_t> tax) const {
+    // Now, trace the line of descent, updating the candidate as we go.
+    GetMRCA();
+
+    int depth = 0;
+    Ptr<taxon_t> test_taxon = tax->GetParent();
+    while (test_taxon) {
+      depth++;
+      if (test_taxon == mrca || !test_taxon) {
+        return depth;
+      }
+      test_taxon = test_taxon->GetParent();
+    }
+    return depth;
+  }
+
+  template <typename ORG, typename ORG_INFO, typename DATA_STRUCT>
+  std::set<Ptr<typename Systematics<ORG, ORG_INFO, DATA_STRUCT>::taxon_t>> Systematics<ORG, ORG_INFO, DATA_STRUCT>::GetCanopyExtantRoots(int time_point) const {
+    // NOTE: This could be made faster by doing something similar to the pairwise distance
+    // function
+    using taxon_t = Systematics<ORG, ORG_INFO, DATA_STRUCT>::taxon_t;
+    std::set< Ptr<taxon_t>> result;
+    // std::cout << "starting " << time_point << std::endl;
+    for (Ptr<taxon_t> tax : active_taxa) {
+        // std::cout << tax->GetInfo() << std::endl;
+      while (tax) {
+        // std::cout << tax->GetInfo() << " " << tax->GetOriginationTime() << " " << tax->GetDestructionTime() << std::endl;
+        if (tax->GetOriginationTime() <= time_point && tax->GetDestructionTime() > time_point ) {
+          result.insert(tax);
+        // std::cout << "inserting " << tax->GetInfo() << std::endl;
+          break;
+        }
+        tax = tax->GetParent();
+      }
+    }
+
+    return result;
+
+  }
+
+  template <typename ORG, typename ORG_INFO, typename DATA_STRUCT>
+  int Systematics<ORG, ORG_INFO, DATA_STRUCT>::GetPhylogeneticDiversityNormalize(int generation, std::string filename) const {
```

Comment:
If generations are limited to 0, 10, 20, 30, ... 100. we should probably test the generation argument and give an error message if that is violated,

Reply:
Oh sorry I was misrememebing slightly when I left my earlier comment about normalization. This actually does handle generations that aren't increments of 10. It just bins them together